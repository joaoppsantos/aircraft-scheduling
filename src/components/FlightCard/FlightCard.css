body {
    --card-bg-color: #add8e6;
    --hover-bg-color: #95d3e7;
}

.flight-card {
    height: 9vh;
    width: 100%;

    font-size: 17px;
    margin-top: 15px;

    display: flex;
    flex-wrap: wrap;
    background-color: var(--card-bg-color);
    align-items: center;
    flex-direction: column;
}

.flight-card:hover {
    background-color: var(--hover-bg-color);
}

.flight-id {
    position: relative;
    font-size: 12px;
    align-self: flex-start;
    padding: 5px;
}
