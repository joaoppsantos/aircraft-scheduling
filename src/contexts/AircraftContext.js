import { createContext, useEffect, useState } from 'react';
import request from '../request';

export const AircraftContext = createContext({});

export function AircraftProvider({ children }) {
	const [aircrafts, setAircrafts] = useState([]);
	const [isAircraftLoading, setIsAircraftLoading] = useState(true);
	const [currentAircraft, setCurrentAircraft] = useState(undefined);
	const [currentPercentage, setCurrentPercentage] = useState(0);

	const setPercentage = percentage => {
		setCurrentPercentage(percentage);
	};
	const selectAircraft = aircraft => {
		setCurrentAircraft(aircraft);
	};

	useEffect(() => {
		request.get(`/aircrafts?limit=5&offset=0`).then((response) => {
			setAircrafts(response.data.data);
			setIsAircraftLoading(false);
		});
	}, []);

	return (
		<AircraftContext.Provider value={{
			aircrafts,
			isAircraftLoading,
			currentAircraft,
			selectAircraft,
			currentPercentage,
			setPercentage
		}}>
			{children}
		</AircraftContext.Provider>
	)
}
