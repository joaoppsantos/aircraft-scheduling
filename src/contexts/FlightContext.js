import request from '../request';
import { createContext, useEffect, useState } from 'react';

export function FlightProvider({ children }) {
	const [flights, setFlights] = useState([]);
	const [activeFlights, setActiveFlights] = useState([]);
	const [offset, setOffset] = useState(-1);
	const [hasFlightsRemaining, setHasFlightsRemaining] = useState(true);
	const paginationLimit = 25;

	const toggleActiveFlight = (flight) => {
		let isFlightOnList = false;

		activeFlights.forEach((activeInfo) => {
			if (activeInfo.id === flight.id) {
				isFlightOnList = true;
			}
		});

		isFlightOnList ? 
			setActiveFlights(activeFlights.filter(activeInfo => activeInfo.id !== flight.id))
			: setActiveFlights([...activeFlights, flight]);
	}

	useEffect(() => {
		if (!hasFlightsRemaining) {
			return;
		}

		async function fetchData() {
			try {
				const newOffset = offset + 1;
				const response = await request.get(`/flights?limit=${paginationLimit}&offset=${newOffset}`);

				setOffset(newOffset);
				setFlights([...flights, ...response?.data?.data]);

				if (paginationLimit * (newOffset + 1) > response?.data?.pagination?.total) {
					setHasFlightsRemaining(false);
				}
			} catch (error) {
				console.log('Failed to load !');
			};
		}

		fetchData();
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [flights, hasFlightsRemaining]);

	return (
		<FlightContext.Provider value={{
			flights,
			toggleActiveFlight,
			activeFlights
		}}>
			{ children }
		</FlightContext.Provider>
	)
}

export const FlightContext = createContext({});