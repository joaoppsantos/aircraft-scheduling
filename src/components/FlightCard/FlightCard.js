import './FlightCard.css';
import { useContext } from 'react';
import { FlightContext } from '../../contexts/FlightContext';

const FlightCard = flight => {
	const { flights, toggleActiveFlight } = useContext(FlightContext);

	const handleCardClick = () => {
		// verificar se o schedule já está ocupado
		// verificar se o aeroporto de destino foi o último
		// verificar se já passaram 20 minutos do último vôo

		

		toggleActiveFlight(flight);
	};

	return (
		<div>
			<div className="flight-card" onClick={handleCardClick}> 
				<span className="flight-id">{flight?.id}</span>
				<span>{flight?.origin} {'>'} {flight?.destination}</span>
				<span>{flight?.readable_departure} {'>'} {flight?.readable_arrival}</span>
			</div>
		</div>
	)
};

export default FlightCard;
