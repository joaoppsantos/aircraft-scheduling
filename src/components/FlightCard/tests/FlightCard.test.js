import * as React from 'react';
import { render } from '@testing-library/react';
import FlightCard from '../FlightCard';
import { FlightContext } from '../../../contexts/FlightContext';

const mockFlightCardProps = {
    "id": "AS1001",
    "departuretime": 21600,
    "arrivaltime": 26100,
    "readable_departure": "06:00",
    "readable_arrival": "07:15",
    "origin": "LFSB",
    "destination": "LFMN"
};

describe('<FlightCard />', () => {
    it('should render correctly', () => {
        const mockFlightContext = {
            activeFlights: [
                mockFlightCardProps,
            ],
            toggleActiveFlight: jest.fn(),
        };

        const { container } = render(
            <FlightContext.Provider value={mockFlightContext}>
                <FlightCard {...mockFlightCardProps} />
            </FlightContext.Provider>
        );

        expect(container).toMatchSnapshot();
    });
});
