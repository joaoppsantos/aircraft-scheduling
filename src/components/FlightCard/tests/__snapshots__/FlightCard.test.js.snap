// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlightCard /> should render correctly 1`] = `
<div>
  <div>
    <div
      class="flight-card"
    >
      <span
        class="flight-id"
      >
        AS1001
      </span>
      <span>
        LFSB
         
        &gt;
         
        LFMN
      </span>
      <span>
        06:00
         
        &gt;
         
        07:15
      </span>
    </div>
  </div>
</div>
`;
